lib/evmone/eof.cpp

```diff
@@ -90,37 +151,279 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+        std::accumulate(
+            section_headers[CODE_SECTION].begin(), section_headers[CODE_SECTION].end(), 0) +
+        (!section_headers[DATA_SECTION].empty() ? section_headers[DATA_SECTION].front() : 0);
     const auto remaining_container_size = container_end - it;
     if (section_bodies_size != remaining_container_size)
         return EOFValidationError::invalid_section_bodies_size;
 
+    if (!section_headers[TYPE_SECTION].empty() &&
+        section_headers[TYPE_SECTION][0] != section_headers[CODE_SECTION].size() * 4)
+        return EOFValidationError::invalid_type_section_size;
+
     return section_headers;
 }
 
+std::variant<std::vector<EOFCodeType>, EOFValidationError> validate_types(
+    bytes_view container, size_t header_size, std::vector<uint16_t> type_section_sizes) noexcept
+{
+    assert(!container.empty());              // guaranteed by EOF headers validation
+    assert(type_section_sizes.size() <= 1);  // guaranteed by EOF headers validation
+
+    if (type_section_sizes.empty())
+        return std::vector{EOFCodeType{0, 0, 0}};
+
+    std::vector<EOFCodeType> types;
+
+    // guaranteed by EOF headers validation
+    assert(header_size + type_section_sizes[0] < container.size());
+
+    for (auto offset = header_size; offset < header_size + type_section_sizes[0]; offset += 4)
+    {
+        types.emplace_back(
+            container[offset], container[offset + 1], read_uint16_be(&container[offset + 2]));
+    }
+
+    // check 1st section is (0, 0)
+    if (types[0].inputs != 0 || types[0].outputs != 0)
+        return EOFValidationError::invalid_first_section_type;
+
+    for (const auto& t : types)
+    {
+        if (t.outputs > OUTPUTS_INPUTS_NUMBER_LIMIT || t.inputs > OUTPUTS_INPUTS_NUMBER_LIMIT)
+            return EOFValidationError::inputs_outputs_num_above_limit;
+
+        if (t.max_stack_height > MAX_STACK_HEIGHT)
+            return EOFValidationError::max_stack_height_above_limit;
+    }
+
+    return types;
+}
+
 EOFValidationError validate_instructions(evmc_revision rev, bytes_view code) noexcept
 {
     assert(!code.empty());  // guaranteed by EOF headers validation
 
-    size_t i = 0;
-    uint8_t op = code[0];
-    while (i < code.size())
+    const auto& cost_table = baseline::get_baseline_cost_table(rev, 1);
+
+    for (size_t i = 0; i < code.size(); ++i)
     {
-        op = code[i];
-        const auto& since = instr::traits[op].since;
-        if (!since.has_value() || *since > rev)
+        const auto op = code[i];
+        if (cost_table[op] == instr::undefined)
             return EOFValidationError::undefined_instruction;
 
-        i += instr::traits[op].immediate_size;
-        ++i;
-    }
+        if (op == OP_RJUMPV)
+        {
+            if (i + 1 >= code.size())
+                return EOFValidationError::truncated_instruction;
+
+            const auto count = code[i + 1];
+            if (count < 1)
+                return EOFValidationError::invalid_rjumpv_count;
```

Comment:
new spec is
```suggestion
            const auto count = code[i + 1] + 1;
```
but ok to change it in follow-up PR

Reply:
the change is not merged to the spec yet, so we'll change implementation later